### DISCUSSION TITLE
cobbler problem with hivex

### DISCUSSION THREAD

**Discussion by mitodeboa** [cobbler/cobbler#2621] [Version specific questions]
Hi @SchoolGuy Cobbler sync_post_wingen.py does nothing. I have verified that HAS_HIVEX is false although it does not appear in the log
I have installed python3-hivex-1.3.18-20.module_el8.5.0 + 746 + bbd5d70c.x86_64 but this installs it in python3.6. If I copy it to python39 and try to use it it gets 
<pre>
Python 3.9.2 (default, Mar 29 2021, 14:11:56) 
[GCC 8.4.1 20200928 (Red Hat 8.4.1-1)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import hivex
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib64/python3.9/site-packages/hivex/__init__.py", line 37, in <module>
    import libhivexmod
ModuleNotFoundError: No module named 'libhivexmod'

ll /usr/lib64/python3.9/site-packages/
total 444
drwxr-xr-x 3 root root     44 abr 15 11:37 curl
drwxr-xr-x 3 root root     65 abr 20 12:39 hivex
-rwxr-xr-x 1 root root  25784 abr 20 12:39 libhivexmod.cpython-36m-x86_64-linux-gnu.so
drwxr-xr-x 6 root root     84 abr 19 10:42 mod_wsgi
drwxr-xr-x 2 root root    134 abr 19 10:42 mod_wsgi-4.7.1-py3.9.egg-info
drwxr-xr-x 2 root root      6 mar 29 16:14 __pycache__
-rw-r--r-- 1 root root   4690 ago  4  2020 pycurl-7.43.0.5-py3.9.egg-info
-rwxr-xr-x 1 root root 159624 ago  4  2020 pycurl.cpython-39-x86_64-linux-gnu.so
-rw-r--r-- 1 root root   1781 mar 30 22:46 PyYAML-5.3.1-py3.9.egg-info
-rw-r--r-- 1 root root    119 feb 19 13:31 README.txt
drwxr-xr-x 3 root root    333 abr 15 11:03 yaml
-rwxr-xr-x 1 root root 245832 mar 30 22:46 _yaml.cpython-39-x86_64-linux-gnu.so
</pre>

Regards

**Comment by dansou901** [comment]
Ping  @tpw56j, can you help out here?

**Reply by tpw56j** [reply]
@dansou901 For python 3.9, the ``/usr/lib64/python3.9/site-packages/libhivexmod.cpython-39-x86_64-linux-gnu.so`` file is needed. Maybe compile python3-hivex from sources?

    # rpm -q python3-hivex
    python3-hivex-1.3.19-4.fc33.1.x86_64
    # rpm -ql python3-hivex
    /usr/lib/.build-id
    /usr/lib/.build-id/39
    /usr/lib/.build-id/39/06846855a8816debb1849eb1a593552d83a4b3
    /usr/lib64/python3.9/site-packages/hivex
    /usr/lib64/python3.9/site-packages/hivex/__init__.py
    /usr/lib64/python3.9/site-packages/hivex/__pycache__
    /usr/lib64/python3.9/site-packages/hivex/__pycache__/__init__.cpython-39.opt-1.pyc
    /usr/lib64/python3.9/site-packages/hivex/__pycache__/__init__.cpython-39.pyc
    /usr/lib64/python3.9/site-packages/hivex/__pycache__/hive_types.cpython-39.opt-1.pyc
    /usr/lib64/python3.9/site-packages/hivex/__pycache__/hive_types.cpython-39.pyc
    /usr/lib64/python3.9/site-packages/hivex/hive_types.py
    /usr/lib64/python3.9/site-packages/libhivexmod.cpython-39-x86_64-linux-gnu.so

**Reply by mitodeboa** [reply]
Ok compiled and HAS_HIVEX now is true
Thanks @tpw56j

**Comment by mitodeboa** [comment]
Anyway, the sync_post_wingen.py script still does nothing.  This is the output:
<pre>
running shell triggers from /var/lib/cobbler/triggers/sync/post/*
running shell trigger /var/lib/cobbler/triggers/sync/post/sync_post_wingen.py
running: ['/var/lib/cobbler/triggers/sync/post/sync_post_wingen.py']
Exception occurred: <class 'OSError'>
Exception value: [Errno 8] Exec format error: '/var/lib/cobbler/triggers/sync/post/sync_post_wingen.py'
Exception Info:
Exception occurred: <class 'cobbler.cexceptions.CX'>
Exception value: "OS Error, command not found?  While running: ['/var/lib/cobbler/triggers/sync/post/sync_post_wingen.py']
</pre>
After adding the shebang line (#!/usr/bin/env python) I do not get errors but the script does not make any function calls so it does nothing. 
<pre>
running shell trigger /var/lib/cobbler/triggers/sync/post/sync_post_wingen.py
running: ['/var/lib/cobbler/triggers/sync/post/sync_post_wingen.py']
received on stdout: 
received on stderr: 
shell trigger /var/lib/cobbler/triggers/sync/post/sync_post_wingen.py finished successfully
</pre>
What I have to do? is it necessary to further modify the script to make a call to the "run" function? I guess i'm doing something wrong i shouldn't have to modify it 

I followed the manual https://cobbler.readthedocs.io/en/latest/user-guide/wingen.html?highlight=cobbler%20windows 

[profile_dumpvars.log](https://github.com/cobbler/cobbler/files/6364945/profile_dumpvars.log)

**Reply by tpw56j** [reply]
@mitodeboa Error in the documentation. It has been fixed at https://github.com/cobbler/cobbler/pull/2575, but the PR has not yet been merged into mainstream. There should be no commas in ``-kernel-options``: 

    --kernel-options='pxeboot=win10a.0 bootmgr=boot1ea.exe bcd=1Ea winpe=winpe.wim sif=autounattended.xml'

**Reply by mitodeboa** [reply]
I have removed the commas and it still does not generate the files sync_post_wingen.py does nothing. If I put a print in the run function, nothing comes out of stdout. The run function, which is the one that generates the files, is not being invoked

**Reply by tpw56j** [reply]
@mitodeboa Please check the installed packages, distro, profile and the existence of the required Windows files, or share the output of the commands below:

    # rpm -q cobbler python3-hivex python3-pefile wimlib-utils
    cobbler-3.2.0-3.fc34.noarch
    python3-hivex-1.3.19-8.fc34.x86_64
    python3-pefile-2019.4.18-5.fc34.noarch
    wimlib-utils-1.13.3-2.fc34.x86_64

    # ls -l /var/lib/tftpboot/winos/{winpe8.template,Win2016-Server_EN-x64/{pxeboot.n12,bootmgr.exe,boot/{BCD,boot.sdi}}}
    -r-xr-xr-x. 1 nobody nobody     16384 Jul 16  2016 /var/lib/tftpboot/winos/Win2016-Server_EN-x64/boot/BCD
    -r-xr-xr-x. 1 nobody nobody   3170304 May 26  2016 /var/lib/tftpboot/winos/Win2016-Server_EN-x64/boot/boot.sdi
    -rwxrwxrwx. 1 root   root      700256 Aug 22  2013 /var/lib/tftpboot/winos/Win2016-Server_EN-x64/bootmgr.exe
    -rwxrwxrwx. 1 root   root       25358 Aug 22  2013 /var/lib/tftpboot/winos/Win2016-Server_EN-x64/pxeboot.n12
    -rw-rw-r--. 1 root   root   209074255 May 17  2019 /var/lib/tftpboot/winos/winpe8.template

    # cobbler version
    Cobbler 3.2.0
      source: ?, ?
      build time: Tue Jan 26 00:00:00 2021

    # cobbler distro report --name=Win2016-Server_EN-x64
    Name                           : Win2016-Server_EN-x64
    Architecture                   : x86_64
    Automatic Installation Template Metadata : {}
    TFTP Boot Files                : {}
    Boot loader                    : <<inherit>>
    Breed                          : windows
    Comment                        : 
    Fetchable Files                : {}
    Initrd                         : /var/lib/tftpboot/winos/Win2016-Server_EN-x64/boot/boot.sdi
    Kernel                         : /var/lib/tftpboot/winos/Win2016-Server_EN-x64/pxeboot.n12
    Kernel Options                 : {}
    Kernel Options (Post Install)  : {}
    Management Classes             : []
    OS Version                     : 2016
    Owners                         : ['admin']
    Redhat Management Key          : <<inherit>>
    Remote Boot Initrd             : ~
    Remote Boot Kernel             : ~
    Template Files                 : {}

    # cobbler profile report --name=Win2016-Server_EN-x64
    Name                           : Win2016-Server_EN-x64
    Automatic Installation Template : win.ks
    Automatic Installation Metadata : {}
    TFTP Boot Files                : {}
    Comment                        : 
    DHCP Tag                       : default
    Distribution                   : Win2016-Server_EN-x64
    Enable gPXE?                   : 0
    Enable PXE Menu?               : False
    Fetchable Files                : {}
    DHCP Filename Override         : <<inherit>>
    Kernel Options                 : {'bcd': '6Ea', 'winpe': 'winpe.wim', 'bootmgr': 'boot6ea.exe', 'pxeboot': win10a.0', 'sif': 'autounattended.xml'}
    Kernel Options (Post Install)  : {}
    Management Classes             : []
    Management Parameters          : <<inherit>>
    Name Servers                   : []
    Name Servers Search Path       : []
    Next Server Override           : <<inherit>>
    Owners                         : ['admin']
    Parent Profile                 : 
    Proxy                          : 
    Red Hat Management Key         : <<inherit>>
    Repos                          : []
    Server Override                : <<inherit>>
    Template Files                 : {}
    Virt Auto Boot                 : 1
    Virt Bridge                    : virbr0
    Virt CPUs                      : 1
    Virt Disk Driver Type          : raw
    Virt File Size(GB)             : 5
    Virt Path                      : 
    Virt RAM (MB)                  : 512
    Virt Type                      : qemu


If everything is ok, the ``cobbler sync`` command should run the trigger and create the necessary files: 

    # cobbler sync
    task started: 2021-04-27_102508_sync
    task started (id=Sync, time=Tue Apr 27 10:25:08 2021)
    <...>
    running post-sync triggers
    running python triggers from /var/lib/cobbler/triggers/sync/post/*
    running python trigger cobbler.modules.sync_post_wingen

    Windows profiles:
    Profile: Win2016-Server_EN-x64
    Build answer file: /var/lib/tftpboot/winos/Win2016-Server_EN-x64/sources/autounattended.xml
    Build PXEBoot: /var/lib/tftpboot/winos/Win2016-Server_EN-x64/win10a.0
    Build Loader: /var/lib/tftpboot/winos/Win2016-Server_EN-x64/boot6ea.exe
    Build BCD: /var/lib/tftpboot/winos/Win2016-Server_EN-x64/boot/6Ea for /winos/Win2016-Server_EN-x64/boot/winpe.wim
    running: /usr/bin/cp --reflink=auto /var/lib/tftpboot/winos/winpe8.template /var/lib/tftpboot/winos/Win2016-Server_EN-x64/boot/winpe.wim
    received on stdout: 
    received on stderr: 
    running: /usr/bin/wimupdate /var/lib/tftpboot/winos/Win2016-Server_EN-x64/boot/winpe.wim --command="add /tmp/tmp57ne91hn /Windows/System32/startnet.cmd"
    received on stdout: Scanning "/tmp/tmp57ne91hn" (loading as WIM path: "/Windows/System32/startnet.cmd")...

    1278 bytes scanned (1 files, 0 directories)    
    1278 bytes scanned (1 files, 0 directories)    
    Updating "/Windows/System32/startnet.cmd" in WIM image
    Using LZX compression with 1 thread

    Archiving file data: 0 bytes of 1278 bytes (0%) done
    Archiving file data: 1278 bytes of 1278 bytes (100%) done

**Comment by mitodeboa** [comment]
Hi @tpw56j thanks for the answer.
My intention was to be able to do it in centos 8 stream but of course I do not have those packages available so I have built the version of cobbler 3.2.1, python3-hivex I have also built it and to finish before python3-pefile-2019.4.18-4 .fc33.noarch I have found it at https://download-ib01.fedoraproject.org/pub/fedora/linux/releases/33/Everything/x86_64/os/ and wimlib-utils I have found it as wimtools-1.9.2- 2.el7.nux.x86_64 at http://li.nux.ro/download/nux/dextop/el7/x86_64/nux-dextop-release-0-5.el7.nux.noarch.rpm.
Anyway even though the packages don't match I suppose I should see errors

**Reply by tpw56j** [reply]
@mitodeboa All packages on centos 8 except wimlib-utils can be installed from public repos (EPEL + Oracle Linux).

    # cat /etc/redhat-release 
    Red Hat Enterprise Linux release 8.3 (Ootpa)

    # rpm -q python36
    python36-3.6.8-2.module+el8.3.0+7694+550a8252.x86_64

    # rpm -q cobbler
    cobbler-3.2.1-1.el8.noarch

    # yum repolist
    repo id                                repo name
    appstream                              CentOS Linux 8 - AppStream
    baseos                                 CentOS Linux 8 - BaseOS
    epel                                   Extra Packages for Enterprise Linux 8 - x86_64
    epel-modular                           Extra Packages for Enterprise Linux Modular 8 - x86_64
    extras                                 CentOS Linux 8 - Extras
    ol8_appstream                          Oracle Linux 8 Application Stream (x86_64)
    ol8_baseos_latest                      Oracle Linux 8 BaseOS Latest (x86_64)
    powertools                             CentOS Linux 8 - PowerTools

    # yum install python3-hivex python3-pefile
    Dependencies resolved.
    ==============================================================================================================================================================================================================================================
     Package                                               Architecture                                  Version                                                                       Repository                                            Size
    ==============================================================================================================================================================================================================================================
    Installing:
     python3-hivex                                         x86_64                                        1.3.18-20.module+el8.3.0+7860+a7792d29                                        ol8_appstream                                         35 k
     python3-pefile                                        noarch                                        2019.4.18-5.el8                                                               epel                                                 114 k
    Installing dependencies:
     hivex                                                 x86_64                                        1.3.18-20.module+el8.3.0+7860+a7792d29                                        ol8_appstream                                        112 k

    Transaction Summary
    ==============================================================================================================================================================================================================================================
    Install  3 Packages

    Total download size: 261 k
    Installed size: 718 k
    Is this ok [y/N]: y

I downloaded the source code for wimlib from the official site https://wimlib.net. wimlib-1.13.4.tar.gz lacks the wimtools.spec file to build the rpm, so I took it from wimlib-1.13.1.tar.gz and replaced the version number in it.

    # rpm -q wimtools
    wimtools-1.13.4-1.x86_64

Thereafter:

    # cobbler sync
    task started: 2021-04-29_213853_sync
    task started (id=Sync, time=Thu Apr 29 21:38:53 2021)
    <...>
    Profile: Win2016-Server_EN-x64
    Build answer file: /var/lib/tftpboot/winos/Win2016-Server_EN-x64/sources/autounattended.xml
    Build PXEBoot: /var/lib/tftpboot/winos/Win2016-Server_EN-x64/win10a.0
    Build Loader: /var/lib/tftpboot/winos/Win2016-Server_EN-x64/boot6ea.exe
    Build BCD: /var/lib/tftpboot/winos/Win2016-Server_EN-x64/boot/6Ea for /winos/Win2016-Server_EN-x64/boot/winpe.wim
    running: /usr/bin/cp --reflink=auto /var/lib/tftpboot/winos/winpe8.template /var/lib/tftpboot/winos/Win2016-Server_EN-x64/boot/winpe.wim
    received on stdout: 
    received on stderr: 
    running: /usr/bin/wimupdate /var/lib/tftpboot/winos/Win2016-Server_EN-x64/boot/winpe.wim --command="add /tmp/tmp3xh6da_s /Windows/System32/startnet.cmd"
    received on stdout: Scanning "/tmp/tmp3xh6da_s" (loading as WIM path: "/Windows/System32/startnet.cmd")...

**Reply by mitodeboa** [reply]
Thanks a lot @tpw56j. I'm going to start from scratch with your same versions because, it's a long story, but my centos 8 stream running _dnf --enablerepo=epel install cobbler_ only provided me cobbler 3.2.0-2.module_el8+10474+71f42bad from epel-modular that's why i built 3.2.1 and when I did it with python36 I had this issue https://github.com/cobbler/cobbler/issues/2614# and as I saw, in the documentation, that python39 was used so I did it and the issue is gone and then that's why i built hivex, ... 
<pre>
#dnf repolist 
repo id                                     repo name
appstream                               CentOS Stream 8 - AppStream
baseos                                    CentOS Stream 8 - BaseOS
epel                                         Extra Packages for Enterprise Linux 8 - x86_64
epel-modular                           Extra Packages for Enterprise Linux Modular 8 - x86_64
extras                                      CentOS Stream 8 - Extras
powertools                              CentOS Stream 8 - PowerTools

# cobbler --version
Cobbler 3.2.1
  source: d6fa0b14, Fri Apr 9 14:07:36 2021 +0200
  build time: Mon Apr 19 09:05:47 2021

# python --version
Python 3.9.2

# python -c "import hivex;print(hivex.__version__)"
1.3.19

# pip freeze|grep pefile
pefile==2019.4.18

# wimupdate --version
wimlib-imagex (distributed with wimlib 1.9.2)
Copyright (C) 2012-2016 Eric Biggers
License GPLv3+; GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Report bugs to ebiggers3@gmail.com.
</pre>

**Comment by SchoolGuy** [comment]
@mitodeboa Was your problem solved in the meantime? :)
